File: modules/proofmanager/android/src/main/java/expo/modules/proofmanager/ProofManagerModule.kt
package expo.modules.proofmanager

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

class ProofManagerModule : Module() {
    
    // Adding our library
    companion object {
        init {
            try {
                System.loadLibrary("proofmanager")
                println("Successfully loaded native library")
            } catch (e: UnsatisfiedLinkError) {
                println("Failed to load native library: ${e.message}")
                e.printStackTrace()
            }
        }
    }

    // Native function declarations matching Rust implementations
    private external fun generateKeysNative(seedPhrase: String): Map<String, ByteArray>
    
    private external fun generateAddressNative(
        spendKey: ByteArray,
        index: Int
    ): Map<String, ByteArray>

    private external fun createNoteNative(
        debtorAddress: Map<String, ByteArray>,
        creditorAddress: Map<String, ByteArray>,
        amount: Long,
        assetId: Long
    ): Map<String, ByteArray>

    private external fun signNoteNative(
        seedPhrase: String,
        note: Map<String, ByteArray>
    ): Map<String, ByteArray>
    
    private external fun verifySignatureNative(
        verificationKey: ByteArray,
        commitment: ByteArray,
        signature: ByteArray
    ): Boolean

    override fun definition() = ModuleDefinition {
        Name("ProofManager")

        AsyncFunction("generateKeys") { seedPhrase: String, promise: Promise ->
            try {
                val result = generateKeysNative(seedPhrase)
                promise.resolve(mapOf(
                    "spendKey" to (result["spendKey"]?.map { it.toInt() and 0xFF } ?: listOf()),
                    "viewKey" to (result["viewKey"]?.map { it.toInt() and 0xFF } ?: listOf())
                ))
            } catch (e: Exception) {
                promise.reject("PROOF_ERROR", e.message, e)
            }
        }

        AsyncFunction("generateAddress") { args: Map<String, Any>, promise: Promise ->
            try {
                val spendKey = (args["spendKey"] as? List<Int>)?.map { it.toByte() }?.toByteArray()
                    ?: throw IllegalArgumentException("Invalid spendKey")
                val index = (args["index"] as? Number)?.toInt()
                    ?: throw IllegalArgumentException("Invalid index")

                val result = generateAddressNative(spendKey, index)
                promise.resolve(mapOf(
                    "diversifier" to (result["diversifier"]?.map { it.toInt() and 0xFF } ?: listOf()),
                    "transmissionKey" to (result["transmissionKey"]?.map { it.toInt() and 0xFF } ?: listOf()),
                    "clueKey" to (result["clueKey"]?.map { it.toInt() and 0xFF } ?: listOf())
                ))
            } catch (e: Exception) {
                promise.reject("PROOF_ERROR", e.message, e)
            }
        }

        AsyncFunction("createNote") { args: Map<String, Any>, promise: Promise ->
            try {
                fun convertAddress(address: Map<*, *>): Map<String, ByteArray> {
                    return mapOf(
                        "diversifier" to ((address["diversifier"] as? List<Int>)?.map { it.toByte() }?.toByteArray()
                            ?: throw IllegalArgumentException("Invalid diversifier")),
                        "transmissionKey" to ((address["transmissionKey"] as? List<Int>)?.map { it.toByte() }?.toByteArray()
                            ?: throw IllegalArgumentException("Invalid transmissionKey")),
                        "clueKey" to ((address["clueKey"] as? List<Int>)?.map { it.toByte() }?.toByteArray()
                            ?: throw IllegalArgumentException("Invalid clueKey"))
                    )
                }

                val debtorAddress = (args["debtorAddress"] as? Map<*, *>)?.let { convertAddress(it) }
                    ?: throw IllegalArgumentException("Invalid debtorAddress")
                val creditorAddress = (args["creditorAddress"] as? Map<*, *>)?.let { convertAddress(it) }
                    ?: throw IllegalArgumentException("Invalid creditorAddress")
                val amount = (args["amount"] as? Number)?.toLong()
                    ?: throw IllegalArgumentException("Invalid amount")
                val assetId = (args["assetId"] as? Number)?.toLong()
                    ?: throw IllegalArgumentException("Invalid assetId")

                val result = createNoteNative(debtorAddress, creditorAddress, amount, assetId)
                promise.resolve(mapOf(
                    "commitment" to (result["commitment"]?.map { it.toInt() and 0xFF } ?: listOf()),
                    "debtorAddress" to mapOf(
                        "diversifier" to (result["debtorDiversifier"]?.map { it.toInt() and 0xFF } ?: listOf()),
                        "transmissionKey" to (result["debtorTransmissionKey"]?.map { it.toInt() and 0xFF } ?: listOf()),
                        "clueKey" to (result["debtorClueKey"]?.map { it.toInt() and 0xFF } ?: listOf())
                    ),
                    "creditorAddress" to mapOf(
                        "diversifier" to (result["creditorDiversifier"]?.map { it.toInt() and 0xFF } ?: listOf()),
                        "transmissionKey" to (result["creditorTransmissionKey"]?.map { it.toInt() and 0xFF } ?: listOf()),
                        "clueKey" to (result["creditorClueKey"]?.map { it.toInt() and 0xFF } ?: listOf())
                    )
                ))
            } catch (e: Exception) {
                promise.reject("PROOF_ERROR", e.message, e)
            }
        }

        AsyncFunction("signNote") { args: Map<String, Any>, promise: Promise ->
            try {
                val seedPhrase = args["seedPhrase"] as? String
                    ?: throw IllegalArgumentException("Invalid seedPhrase")
                    
                val noteMap = args["note"] as? Map<String, Any>
                    ?: throw IllegalArgumentException("Invalid note structure")
                
                // Convert the input note to the format expected by native code
                val processedNote = mutableMapOf<String, ByteArray>()
                
                // Extract commitment
                val commitment = (noteMap["commitment"] as? List<Int>)?.map { it.toByte() }?.toByteArray()
                    ?: throw IllegalArgumentException("Invalid commitment in note")
                processedNote["commitment"] = commitment

                // Extract debtor address
                val debtorAddress = noteMap["debtorAddress"] as? Map<*, *>
                    ?: throw IllegalArgumentException("Invalid debtorAddress")
                
                processedNote["debtorDiversifier"] = (debtorAddress["diversifier"] as? List<Int>)
                    ?.map { it.toByte() }?.toByteArray()
                    ?: throw IllegalArgumentException("Invalid debtor diversifier")
                processedNote["debtorTransmissionKey"] = (debtorAddress["transmissionKey"] as? List<Int>)
                    ?.map { it.toByte() }?.toByteArray()
                    ?: throw IllegalArgumentException("Invalid debtor transmissionKey")
                processedNote["debtorClueKey"] = (debtorAddress["clueKey"] as? List<Int>)
                    ?.map { it.toByte() }?.toByteArray()
                    ?: throw IllegalArgumentException("Invalid debtor clueKey")

                // Extract creditor address
                val creditorAddress = noteMap["creditorAddress"] as? Map<*, *>
                    ?: throw IllegalArgumentException("Invalid creditorAddress")
                
                processedNote["creditorDiversifier"] = (creditorAddress["diversifier"] as? List<Int>)
                    ?.map { it.toByte() }?.toByteArray()
                    ?: throw IllegalArgumentException("Invalid creditor diversifier")
                processedNote["creditorTransmissionKey"] = (creditorAddress["transmissionKey"] as? List<Int>)
                    ?.map { it.toByte() }?.toByteArray()
                    ?: throw IllegalArgumentException("Invalid creditor transmissionKey")
                processedNote["creditorClueKey"] = (creditorAddress["clueKey"] as? List<Int>)
                    ?.map { it.toByte() }?.toByteArray()
                    ?: throw IllegalArgumentException("Invalid creditor clueKey")

                val result = signNoteNative(seedPhrase, processedNote)
                
                // Convert the native result back to the format expected by TypeScript
                promise.resolve(mapOf(
                    "signature" to (result["signature"]?.map { it.toInt() and 0xFF } ?: listOf()),
                    "verificationKey" to (result["verificationKey"]?.map { it.toInt() and 0xFF } ?: listOf()),
                    "note" to mapOf(
                        "commitment" to (result["noteCommitment"]?.map { it.toInt() and 0xFF } ?: listOf()),
                        "debtorAddress" to mapOf(
                            "diversifier" to (result["debtorDiversifier"]?.map { it.toInt() and 0xFF } ?: listOf()),
                            "transmissionKey" to (result["debtorTransmissionKey"]?.map { it.toInt() and 0xFF } ?: listOf()),
                            "clueKey" to (result["debtorClueKey"]?.map { it.toInt() and 0xFF } ?: listOf())
                        ),
                        "creditorAddress" to mapOf(
                            "diversifier" to (result["creditorDiversifier"]?.map { it.toInt() and 0xFF } ?: listOf()),
                            "transmissionKey" to (result["creditorTransmissionKey"]?.map { it.toInt() and 0xFF } ?: listOf()),
                            "clueKey" to (result["creditorClueKey"]?.map { it.toInt() and 0xFF } ?: listOf())
                        )
                    )
                ))
            } catch (e: Exception) {
                println("Error in signNote: ${e.message}")
                e.printStackTrace()
                promise.reject("PROOF_ERROR", "Sign note failed: ${e.message}", e)
            }
        }

        AsyncFunction("verifySignature") { args: Map<String, Any>, promise: Promise ->
            try {
                val verificationKey = (args["verificationKey"] as? List<Int>)?.map { it.toByte() }?.toByteArray()
                    ?: throw IllegalArgumentException("Invalid verificationKey")
                val commitment = (args["commitment"] as? List<Int>)?.map { it.toByte() }?.toByteArray()
                    ?: throw IllegalArgumentException("Invalid commitment")
                val signature = (args["signature"] as? List<Int>)?.map { it.toByte() }?.toByteArray()
                    ?: throw IllegalArgumentException("Invalid signature")

                val result = verifySignatureNative(verificationKey, commitment, signature)
                promise.resolve(result)
            } catch (e: Exception) {
                promise.reject("PROOF_ERROR", e.message, e)
            }
        }
    }
}

File: modules/proofmanager/ios/ProofManagerModule.swift
import ExpoModulesCore

public class ProofManagerModule: Module {


    // Initialize ProofManager instance
    private var proofManager: ProofManager?
    
    public func definition() -> ModuleDefinition {
        Name("ProofManager")
        
        OnCreate {
            do {
                self.proofManager = try ProofManager()
            } catch {
                print("Failed to initialize ProofManager: \(error)")
            }
        }
        
        AsyncFunction("generateKeys") { (seedPhrase: String, promise: Promise) in
            guard let pm = self.proofManager else {
                promise.reject(NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "ProofManager not initialized"]))
                return
            }
            
            do {
                let result = try pm.generateKeys(seedPhrase: seedPhrase)
                promise.resolve([
                    "spendKey": Array(result.spendKey),
                    "viewKey": Array(result.viewKey)
                ])
            } catch {
                promise.reject(NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: error.localizedDescription]))
            }
        }
        
        AsyncFunction("generateAddress") { (args: [String: Any], promise: Promise) in
            guard let pm = self.proofManager else {
                promise.reject(NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "ProofManager not initialized"]))
                return
            }
            
            guard let spendKeyArray = args["spendKey"] as? [Int],
                  let index = args["index"] as? Int else {
                promise.reject(NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid arguments"]))
                return
            }
            
            do {
                let spendKey = Data(spendKeyArray.map { UInt8($0) })
                let result = try pm.generateAddress(spendKeyBytes: spendKey, index: UInt32(index))
                
                promise.resolve([
                    "diversifier": Array(result.diversifier),
                    "transmissionKey": Array(result.transmissionKey),
                    "clueKey": Array(result.clueKey)
                ])
            } catch {
                promise.reject(NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: error.localizedDescription]))
            }
        }
        
        AsyncFunction("createNote") { (args: [String: Any], promise: Promise) in
            guard let pm = self.proofManager else {
                promise.reject(NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "ProofManager not initialized"]))
                return
            }
            
            do {
                // Convert addresses from JS format
                func convertAddress(_ addressDict: [String: Any]) throws -> AddressData {
                    guard let diversifier = addressDict["diversifier"] as? [Int],
                          let transmissionKey = addressDict["transmissionKey"] as? [Int],
                          let clueKey = addressDict["clueKey"] as? [Int] else {
                        throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid address format"])
                    }
                    
                    return AddressData(
                        diversifier: Data(diversifier.map { UInt8($0) }),
                        transmissionKey: Data(transmissionKey.map { UInt8($0) }),
                        clueKey: Data(clueKey.map { UInt8($0) })
                    )
                }
                
                guard let debtorAddressDict = args["debtorAddress"] as? [String: Any],
                      let creditorAddressDict = args["creditorAddress"] as? [String: Any],
                      let amount = args["amount"] as? Int,
                      let assetId = args["assetId"] as? Int else {
                    throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid arguments"])
                }
                
                let debtorAddress = try convertAddress(debtorAddressDict)
                let creditorAddress = try convertAddress(creditorAddressDict)
                
                let result = try pm.createNote(
                    debtorAddress: debtorAddress,
                    creditorAddress: creditorAddress,
                    amount: UInt64(amount),
                    assetId: UInt64(assetId)
                )
                
                promise.resolve([
                    "commitment": Array(result.commitment),
                    "debtorAddress": [
                        "diversifier": Array(result.debtorAddress.diversifier),
                        "transmissionKey": Array(result.debtorAddress.transmissionKey),
                        "clueKey": Array(result.debtorAddress.clueKey)
                    ],
                    "creditorAddress": [
                        "diversifier": Array(result.creditorAddress.diversifier),
                        "transmissionKey": Array(result.creditorAddress.transmissionKey),
                        "clueKey": Array(result.creditorAddress.clueKey)
                      ],
                        "amount": Int(amount),   // <== Added
                        "assetId": Int(assetId)   // <== Added
                    ])
            } catch {
                promise.reject(NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: error.localizedDescription]))
            }
        }
        
        AsyncFunction("signNote") { (args: [String: Any], promise: Promise) in
            guard let pm = self.proofManager else {
                promise.reject(NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "ProofManager not initialized"]))
                return
            }
            
            do {
                guard let seedPhrase = args["seedPhrase"] as? String,
                      let noteDict = args["note"] as? [String: Any] else {
                    throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid arguments"])
                }
                
                // Convert note from JS format
                func convertNote(_ noteDict: [String: Any]) throws -> Note {
                    guard let commitment = (noteDict["commitment"] as? [Int])?.map({ UInt8($0) }),
                          let debtorAddressDict = noteDict["debtorAddress"] as? [String: Any],
                          let creditorAddressDict = noteDict["creditorAddress"] as? [String: Any],
                          let amount = noteDict["amount"] as? Int,
                          let assetId = noteDict["assetId"] as? Int else {
                        throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid note format"])
                    }
                    
                    func convertAddress(_ addressDict: [String: Any]) throws -> AddressData {
                        guard let diversifier = (addressDict["diversifier"] as? [Int])?.map({ UInt8($0) }),
                              let transmissionKey = (addressDict["transmissionKey"] as? [Int])?.map({ UInt8($0) }),
                              let clueKey = (addressDict["clueKey"] as? [Int])?.map({ UInt8($0) }) else {
                            throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid address format"])
                        }
                        
                        return AddressData(
                            diversifier: Data(diversifier),
                            transmissionKey: Data(transmissionKey),
                            clueKey: Data(clueKey)
                        )
                    }
                    
                    return Note(
                        debtorAddress: try convertAddress(debtorAddressDict),
                        creditorAddress: try convertAddress(creditorAddressDict),
                        amount: UInt64(amount),
                        assetId: UInt64(assetId),
                        commitment: Data(commitment)
                    )
                }
                
                let note = try convertNote(noteDict)
                let result = try pm.signNote(seedPhrase: seedPhrase, note: note)
                
                promise.resolve([
                    "signature": Array(result.signature),
                    "verificationKey": Array(result.verificationKey),
                    "note": [
                        "commitment": Array(result.note.commitment),
                        "debtorAddress": [
                            "diversifier": Array(result.note.debtorAddress.diversifier),
                            "transmissionKey": Array(result.note.debtorAddress.transmissionKey),
                            "clueKey": Array(result.note.debtorAddress.clueKey)
                        ],
                        "creditorAddress": [
                            "diversifier": Array(result.note.creditorAddress.diversifier),
                            "transmissionKey": Array(result.note.creditorAddress.transmissionKey),
                            "clueKey": Array(result.note.creditorAddress.clueKey)
                        ]
                    ]
                ])
            } catch {
                promise.reject(NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: error.localizedDescription]))
            }
        }
        
        AsyncFunction("verifySignature") { (args: [String: Any], promise: Promise) in
            guard let pm = self.proofManager else {
                promise.reject(NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "ProofManager not initialized"]))
                return
            }
            
            do {
                guard let verificationKey = (args["verificationKey"] as? [Int])?.map({ UInt8($0) }),
                      let commitment = (args["commitment"] as? [Int])?.map({ UInt8($0) }),
                      let signature = (args["signature"] as? [Int])?.map({ UInt8($0) }) else {
                    throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid arguments"])
                }
                
                let result = try pm.verifySignature(
                    verificationKeyBytes: Data(verificationKey),
                    commitment: Data(commitment),
                    signature: Data(signature)
                )
                
                promise.resolve(result)
            } catch {
                promise.reject(NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: error.localizedDescription]))
            }
        }
        
        AsyncFunction("createIntentAction") { (args: [String: Any], promise: Promise) in
            guard let pm = self.proofManager else {
                promise.reject(NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "ProofManager not initialized"]))
                return
            }
            
            do {
                guard let debtorSeedPhrase = (args["debtorSeedPhrase"] as? [Int])?.map({ UInt8($0) }),
                      let rseedRandomness = (args["rseedRandomness"] as? [Int])?.map({ UInt8($0) }),
                      let debtorIndex = args["debtorIndex"] as? Int,
                      let creditorAddr = args["creditorAddr"] as? String,
                      let amount = args["amount"] as? Int,
                      let assetId = args["assetId"] as? Int else {
                    throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid arguments"])
                }
                
                let result = try pm.createIntentAction(
                    debtorSeedPhrase: Data(debtorSeedPhrase),
                    rseedRandomness: Data(rseedRandomness),
                    debtorIndex: UInt32(debtorIndex),
                    creditorAddr: creditorAddr,
                    amount: UInt64(amount),
                    assetId: UInt64(assetId)
                )
                
                promise.resolve(result)
            } catch {
                promise.reject(NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: error.localizedDescription]))
            }
        }
    }
}